src/gridcoin/mrc.cpp

```diff
@@ -0,0 +1,382 @@
+// Copyright (c) 2014-2022 The Gridcoin developers
+// Distributed under the MIT/X11 software license, see the accompanying
+// file COPYING or https://opensource.org/licenses/mit-license.php.
+
+#include "gridcoin/mrc.h"
+
+#include "amount.h"
+#include "key.h"
+#include "main.h"
+#include "gridcoin/account.h"
+#include "gridcoin/claim.h"
+#include "gridcoin/tally.h"
+#include "gridcoin/beacon.h"
+#include "gridcoin/quorum.h"
+#include "gridcoin/researcher.h"
+#include "util.h"
+#include "wallet/wallet.h"
+
+using namespace GRC;
+
+namespace {
+//!
+//! \brief Get the hash of a subset of the data in the mrc object used as
+//! input to sign or verify a research reward claim.
+//!
+//! \param mrc           MRC to generate a hash for.
+//! \param last_block_hash Hash of the block at the head of the chain.
+//! \param mrc    mrc transaction of the block that contains
+//! the claim.
+//!
+//! \return Hash of the CPID and last block hash contained in the claim.
+//!
+uint256 GetMRCHash(
+    const MRC& mrc,
+    const uint256& last_block_hash)
+{
+    const CpidOption cpid = mrc.m_mining_id.TryCpid();
+
+    if (!cpid) {
+        return uint256();
+    }
+
+        CHashWriter hasher(SER_NETWORK, PROTOCOL_VERSION);
+
+        hasher << *cpid << last_block_hash;
+
+        return hasher.GetHash();
+}
+} // anonymous namespace
+
+// -----------------------------------------------------------------------------
+// Class: MRC
+// -----------------------------------------------------------------------------
+
+MRC::MRC() : MRC(CURRENT_VERSION)
+{
+}
+
+MRC::MRC(uint32_t version)
+    : m_version(version)
+    , m_research_subsidy(0)
+    , m_magnitude(0)
+    , m_magnitude_unit(0)
+{
+}
+
+bool MRC::WellFormed() const
+{
+    if (m_version <= 0 || m_version > MRC::CURRENT_VERSION) {
+        return false;
+    }
+
+    if (!m_mining_id.Valid()) {
+        return false;
+    }
+
+    if (m_client_version.empty()) {
+        return false;
+    }
+
+    if (m_mining_id.Which() == MiningId::Kind::CPID) {
+        if (m_research_subsidy <= 0 || m_signature.empty()) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool MRC::HasResearchReward() const
+{
+    return m_mining_id.Which() == MiningId::Kind::CPID;
```

Comment:
Is this method used?

Reply:
Nope. We can get rid of it. you can't actually submit an MRC in an investor context. This was just carried over when I copied the Claim class.